Tweaks to npm scripts to improve windows compatibility

ajmchambers opened the pull request:
Its currently difficult to built and test lit from a windows machine which makes it difficult to contribute. I've personally ran into issues myself while investigating this issue ( https://github.com/lit/lit/issues/1815 ).

Not sure how best to approach this so apologise if I should have created an issue first, any feedback is welcome. I've went and modified the package.json files for most of the packages (in particular prioritizing everything needed to be able to build and test the lit and lit-labs packages, though currently still having issues with localize and localize-tools).

The goal is to leave everything functioning as before for unix users but have them working on windows. I've ran the build and tests on my mac before and after the changes and the output appears the same so far.

## Main changes:

- replacing `rm -rf` with `rimraf` ( https://github.com/isaacs/rimraf )
- using `cross-env` for setting environment variables ( https://github.com/kentcdodds/cross-env )
- replacing single quoted paths with escaped double quotes e.g. `"npx wtr \"../lit/test/**/*_test.(js|html)\""` instead of `"npx wtr '../lit/test/**/*_test.(js|html)'"`
- in the case of regen-package-lock replacing the `;` AND with `"(rimraf node_modules package-lock.json || true) && npm install"` ( https://medium.com/@chillypenguin/running-node-npm-scripts-sequentially-on-windows-8737dc24da1f )

## Remaining issues:

`lerna run build` from the root fails because the build for `localize` and `localize-tools` is not yet working, so in the meantime I need to build all the other packages individually.

### Issues with building `localize` and `localize-tools`

npm clean script not working on windows, its possible to replace `sed` and `grep` using ( https://github.com/shelljs/shx ) but the `xargs` remains an issue.
```json
 "scripts": {
    "clean": "git clean -dXn | sed 's/^Would remove //' | grep -v node_modules | xargs rm -rf",
  }
```

An alterative is to use `rimraf` and specify the files directly similar to some of the other packages, which would allow the `localize` package to build e.g.:

- for localize: `"rimraf {fnv1a64,id-generation,lit-localize}.{js,js.map,d.ts} tests/ .tsbuildinfo"`
- for localize-tools: `"rimraf lib/"`

However localize-tools has the `src/fnv1a64.ts` and `src/id-generation.ts` files symlinked to the localize counterparts, this a bit trickier for windows as it requires installing git with non-default settings may require running the command with admin privileges, unless enabling a setting in a recent version of windows 10. More info here: ( https://github.com/git-for-windows/git/wiki/Symbolic-Links ). So not sure what to do here yet.

The final thing for localize is the `package.json` files in `localize/examples/runtime` and `localize/examples/transform` which would simply need `rm -rf` replaced with `rimraf`.

-------------------------

The above allows me to build lit on windows, however once built there are some issues with tests (which I have not yet looked into). Here is an example:

![image](https://user-images.githubusercontent.com/5426717/117551225-a5bbb280-b03c-11eb-8285-f3bffff3e0fe.png)

ajmchambers commented:
@googlebot I signed it!

justinfagnani commented:
@aomarks can you check out the localize changes?

ajmchambers commented on packages/localize-tools/package.json:
```diff
@@ -19,13 +19,14 @@
   "type": "module",
   "scripts": {
-    "clean": "git clean -dXn | sed 's/^Would remove //' | grep -v node_modules | xargs rm -rf",
+    "clean": "rimraf lib/",
```
Oops yep you're right here, I was trying to figure out what the output would have been but just misunderstood it.

I believe this would be the correct command `"rimraf {fnv1a64,id-generation,lit-localize}.{js,js.map,d.ts} tests/ .tsbuildinfo examples/runtime/lib/ examples/transform/{bundled/,lib/}"`

Though the original is quite nice, I'll continue to look to see if I can find a similar alternative.

Hmm couple of thoughts though:
- does `git clean -dXn` always output `Would remove ` or are the responses translated to the users language?
- is there any risk of git clean somehow giving an error or unexpected result and end up `rm -rf` files you didn't want to like `*` or something?

ajmchambers replied:
Updated the rimraf version for now.

Was hoping to be able to run the git clean directly like `git clean -dXf --exclude node_modules` but sadly can't get it to work as expected.

Could possibly have a mix of the two e.g.: 
`rimraf tests/ examples/runtime/lib/ examples/transform/{bundled/,lib/} && git clean -Xf`

aomarks replied:
I'm fine with just enumerating the files for now.

> * does `git clean -dXn` always output `Would remove ` or are the responses translated to the users language?

Great question, I don't know. It's also possible for the string to change in English across versions, so it's probably pretty brittle anyway.

> * is there any risk of git clean somehow giving an error or unexpected result and end up `rm -rf` files you didn't want to like `*` or something?

You would have to have added the file you didn't actually want to delete to the `.gitignore`, so I think this would be pretty rare.

aomarks replied:
Actually I just noticed that since https://github.com/lit/lit/pull/1831 has landed, and those symlinks are removed, there's no output for this package in the root, so the simple `rimraf lib/` should work fine.

ajmchambers commented on packages/reactive-element/package.json:
```diff
@@ -45,20 +45,20 @@
   "scripts": {
     "build": "npm run clean && npm run build:ts && rollup -c && npm run build:babel",
     "build:watch": "rollup -c --watch",
-    "build:ts": "tsc --build && treemirror development . '**/*.d.ts{,.map}'",
```
Yea windows and unix treat single quotes differently, think it ends up as `"'quoted text'"` rather than the expected `"quoted text"`, also think double quotes are needed for globs on windows.

In this case the single quoted version gives this error: `'html)'' is not recognised as an internal or external command, operable program or batch file.`

aomarks replied:
TIL thanks!

aomarks commented on packages/localize-tools/src/fnv1a64.ts:
```diff
@@ -1 +0,0 @@
-../../localize/src/fnv1a64.ts
```
Cool! https://github.com/lit/lit/pull/1831 has landed so we should be good.

aomarks commented on packages/localize-tools/src/id-generation.ts:
```diff
@@ -1 +0,0 @@
-../../localize/src/id-generation.ts
```
Fixed in https://github.com/lit/lit/pull/1831

aomarks reviewed (approved):
LGTM thanks for doing this!

aomarks commented:
For the package locks conflicts, you can just run `npm run upgrade` from the root to regenerate everything.

ajmchambers commented:
> For the package locks conflicts, you can just run `npm run upgrade` from the root to regenerate everything.

ah thanks... having a bit of a fight with it

aomarks commented:
Hmm, looks like something is off relating to the package locks being out of sync with the package json files somehow? Not sure why that would happen if you did a full regeneration https://github.com/lit/lit/pull/1867/checks?check_run_id=2570659334

ajmchambers commented:
hmm hopefully sorted now, had been added dependencies via `npx lerna add --dev rimraf --scope packagename` but seems that doesn't update the package-lock.json.

ajmchambers commented:
Hold off on this for now, couple of issues I want to investigate. After running build on windows: 
- git wants to commit some files like `.husky/.gitignore` and `config.schema.json` in localize-tools with crlf file endings
- ignore-sync seems to not be coping across the [relative] section for `.eslintignore` and `.prettierignore`

ajmchambers commented:
Ok I've made some changes and I think we are now good.

- ignore-sync has been updated to 4.0.0 which fixes the [relative] merging issue on windows

### `nuke` and `update` scripts
Updated the `nuke` and `update` scripts to fallback to `npx rimraf` should the `rm -rf` fail:

e.g `nuke`
```
(rm -rf node_modules || npx rimraf node_modules) && npm install && lerna exec \"(rm -rf node_modules || npx rimraf node_modules)\" && lerna bootstrap && npm run clean
```

### Line endings issues

Git for Windows default behaviour is to automatically check out text files with CRLF ending and convert them back to LF when committing.

Running the build was changing some files from CRLF to LF (while the actual content remained the same), which made git think there was a change to commit but gave an error when trying to commit that there was nothing to commit.

Some examples:
- husky updates the `/.husky/.gitignore` and saves it in LF endings
- ignore-sync writes out the generated `.prettierignore`, `.eslintignore` etc with LF endings

To fix this I've added a few `.gitattributes` files, for example in the root:
```
# Auto detect text files and perform LF normalization
* text=auto

# Declare files that will always have LF line endings on checkout.
/.husky/.gitignore text eol=lf
/.prettierignore text eol=lf
/.prettierignore-sync text eol=lf
/.eslintignore text eol=lf
/.eslintignore-sync text eol=lf
```
They have also been added to: `localize-tools`, `localize/examples/runtime` and `localize/examples/transform`.

I've cloned and ran the build on windows and mac and it all seems good.

Also ran the new update script so hopefully my package-lock.json nemesis is taken care of this time.
